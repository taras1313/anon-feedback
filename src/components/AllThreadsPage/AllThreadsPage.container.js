import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { AllThreadsPage as AllThreads } from './AllThreadsPage';
import { getAllThreads } from '../../actions/threadActions';

const mapStateToProps = (
  {
    userReducer: { user: { _id } },
    threadsReducer: { threads }
  }
) => ({ userId: _id, threads });

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators({
    getAllThreads
  }, dispatch)
});

export const AllThreadsPage = connect(mapStateToProps, mapDispatchToProps)(AllThreads);