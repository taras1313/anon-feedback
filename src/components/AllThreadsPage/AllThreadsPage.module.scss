.buttonContainer {
	width: 100%;
	display: flex;
	justify-content: center;
	height: 60px;
	padding: 10px 0px;

	span {
		font-weight: bold;
	}
}

.iconWrapper {
	height: 16px;
	width: 16px;
	margin: 0px 6px;
	svg {
		width: 100%;
		height: 100%;
	}
}

.active {
  color: white !important;
  background-color: #3f51b5 !important;
}
