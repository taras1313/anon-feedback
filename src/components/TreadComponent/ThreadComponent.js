import React, { Component, Fragment, useState } from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import QuestionAnswerOutlinedIcon from '@material-ui/icons/QuestionAnswerOutlined';
import cx from 'classnames';

import styles from './ThreadComponent.module.scss';
import { ThreadAuthorComponent } from '../ThreadAuthorComponent';
import { ThreadAuditComponent } from '../ThreadAuditComponent';
import { ManipulateThreadComponent } from '../ManipulateThreadComponent';
import { CommentComponent } from '../CommentComponent';
import { CreateCommentComponent } from '../CreateCommentComponent';
import Loader from '../Loader/Loader';

const randomize = () => Math.floor(Math.random() * 10);
const PREVIEW = 'preview';
const FULL = 'full';
const viewTypes = [PREVIEW, FULL];

//todo like logic- when to disable, already liked functionality, dislike cancels like etc.

export class ThreadComponent extends Component {
  constructor(props) {
    super(props);
    const { threadView } = props;

    this.renderContent = threadView === FULL ? this.renderThread : this.renderPreviewThread;

    this.state = {
      editThreadOpen: false
    };
  }

  closeEditThread = () => this.setState({ editThreadOpen: false });

  openEditThread = () => {
    const {
      actions: { setThreadData },
      thread: { text, _id, title }
    } = this.props;
    setThreadData({ text, _id, title });

    this.setState({ editThreadOpen: true });
  };

  renderPreviewThread = () => {
    const {
      thread: {
        author: { username } = {},
        answersCount,
        subscribersCount,
        dislikesCount,
        likesCount
      }
    } = this.props;

    return (
      <Fragment>
        <div>
          <ThreadAuthorComponent username={username} />
          {this.renderThreadHeader()}
          <ThreadAuditComponent
            subscribersCount={subscribersCount}
            dislikesCount={dislikesCount}
            likesCount={likesCount}
						likeStatus={this.likeStatus()}
					/>
        </div>
        <div className={styles.answersCount}>
          <div className={styles.counterWrapper}>
            <QuestionAnswerOutlinedIcon />
            {answersCount}
          </div>
          <span>answer(s)</span>
        </div>
      </Fragment>
    );
  };

  subscriptionHandler = () => {
    const {
      actions: {
        subscribeToThread,
        unsubscribeFromThread
      },
      thread: { _id: threadId },
      userId
    } = this.props;

    const isSubscribed = this.isSubscribed();
    const params = { threadId, userId };

    if (isSubscribed) {
      unsubscribeFromThread(params)
    } else {
      subscribeToThread(params);
    }
  };

  isSubscribed = () => {
    const { thread: { subscribers }, userId } = this.props;
    return !!subscribers.find(user => user.userId === userId);
  };

  isAuthor = () => {
    const { thread: { author: { userId: authorId } }, userId } = this.props;

    return authorId === userId;
  };

  findUserInThread = () => {
    const { userId, thread: { users } } = this.props;

    return users.find(user => user.userId === userId);
  };

	likeStatus = () => {
		const {
			userId,
			thread: {
				likesList,
				dislikesList
			}
		} = this.props;

		if (likesList.find(el => el.userId === userId)) return 'liked';
		if (dislikesList.find(el => el.userId === userId)) return 'disliked';
		return null
	};

  renderThread = () => {
    const { editThreadOpen } = this.state;

    const {
      userId,
      thread: {
        text,
        subscribersCount,
        dislikesCount,
        likesCount,
        commentsList,
        author: { userId: threadAuthorId, username: authorName } = {}
      },
			likeThread,
			dislikeThread,
      actions: { onCreateComment, onUpdateComment }
    } = this.props;
    const subscribed = this.isSubscribed();
    const { username } = this.findUserInThread() || {};

    return (
      <div className={styles.threadFullWrapper}>
        <div className={styles.headingWrapper}>
          <ThreadAuthorComponent username={authorName} />
          {
            this.isAuthor() && <Button onClick={this.openEditThread} color="secondary">
              Edit
            </Button>
          }
        </div>
        <div className={styles.divider} />
        {this.renderThreadHeader()}
        <div className={styles.description}>
          {text}
        </div>
        <div className={styles.subsWrapper}>

          <ThreadAuditComponent
            subscribersCount={subscribersCount}
            dislikesCount={dislikesCount}
            likesCount={likesCount}
						likeStatus={this.likeStatus()}
						dislike={dislikeThread.bind(null, {id, userId})}
						like={likeThread.bind(null, {id, userId})}
						isClickable
          />

          <Button
            className={styles.btn}
            variant="outlined"
            onClick={this.subscriptionHandler}
            color={subscribed ? 'secondary' : 'primary'}
          >
            {subscribed ? 'Unsubscribe' : 'Subscribe'}
          </Button>
        </div>
        <div className={styles.dividerWrapper}>
          Total comments ({8})
        </div>

        <div className={styles.comments}>
          {commentsList.map(el => {

            return (
              <CommentComponent
                onUpdateComment={onUpdateComment}
                editable={el.author.userId === userId}
                comment={el}
                isAuthorsComment={el.author.userId === threadAuthorId}
              />
            )
          })}
        </div>

        <div className={styles.dividerWrapper}>
          {!!username ?
            <span className={styles.userHighlight}>{username},</span> : ''} Write your comment here
        </div>
        <CreateCommentComponent onCreateComment={onCreateComment} username={username} />

        <ManipulateThreadComponent
          onClose={this.closeEditThread}
          isOpen={editThreadOpen}
          action="edit"
        />
      </div>
    );
  };

  renderThreadHeader = () => {
    const { threadView } = this.props;
    const { title = 'What to learn SWIFT or GO?', _id: id } = this.props.thread;

    return (
      <NavLink to={`/feed/${id}`}>
        <div
          className={cx(styles.heading, {
            [styles.hovered]: threadView === PREVIEW
          })}
        >
          {title}
        </div>
      </NavLink>
    );
  };


  render() {
    return <div className={styles.wrapper}>{this.renderContent()}</div>;
  }
}

ThreadComponent.propTypes = {
  heading: PropTypes.string,
  answersCount: PropTypes.number,
  subscribersCount: PropTypes.number,
  creationDate: PropTypes.object,
  content: PropTypes.string,
  likesCount: PropTypes.number,
  dislikesCount: PropTypes.number,
  threadView: PropTypes.oneOf(viewTypes),
  author: PropTypes.object,
  onLikeClick: PropTypes.func,
  onDislikeClick: PropTypes.func,
  onSubscribeClick: PropTypes.func,
  onEditClick: PropTypes.func
};

ThreadComponent.defaultProps = {
  threadView: FULL, // edit
  subscribersCount: randomize(),
  answersCount: randomize(),
  dislikesCount: randomize(),
  likesCount: randomize(),
  author: {}
};
