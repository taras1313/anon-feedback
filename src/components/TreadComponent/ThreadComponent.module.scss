@import "../../styles/variables";

.wrapper {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  padding: 30px 0;
  border-bottom: 1px solid $dirty-white;
}

.answersCount {
  display: flex;
  flex-direction: column;
  color: $light-green;
}

.heading {
  font-size: 21px;
  font-weight: bold;
}

.hovered {
  &:hover {
    color: $dark-violet;
    cursor: pointer;
  }
}

.description {
  margin: 16px 0;
  font-size: 15px;
  line-height: 24px;
}

//.threadMetaInfo {
//  color: $light-gray;
//  display: flex;
//  font-size: 13px;
//  width: 100%;
//  align-items: flex-end;
//}

.counterWrapper {
  display: flex;
  justify-content: space-evenly;
  font-weight: bold;
  margin-bottom: 6px;
}

//.divider {
//  margin: 0 8px;
//}
//
//.icon {
//  max-width: 20px;
//  max-height: 20px;
//  margin-right: 6px;
//}