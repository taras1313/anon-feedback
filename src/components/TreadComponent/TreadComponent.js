import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import QuestionAnswerOutlinedIcon from '@material-ui/icons/QuestionAnswerOutlined';
import ThumbDownOutlinedIcon from '@material-ui/icons/ThumbDown';
import ThumbUpOutlinedIcon from '@material-ui/icons/ThumbUp';
import cx from 'classnames';

import styles from './TreadComponent.module.scss';

const randomize = () => Math.floor(Math.random() * 10);
const PREVIEW = 'preview';
const FULL = 'full';
const viewTypes = [PREVIEW, FULL];
//todo like logic- when to disable, already liked functionality, dislike cancels like etc.

export class TreadComponent extends Component {
  constructor(props) {
    super(props);
    const { treadView } = props;

    this.renderContent = treadView === FULL ? this.renderTread : this.renderPreviewTread;
  }

  renderTread = () => {
    const { answersCount } = this.props;

    return (
      <Fragment>
        <div>
          {this.renderTreadAuthor()}
          {this.renderTreadHeader()}
          <div className={styles.description}>
            I want to loot 1.5k but I dunno where to start and which programming language should i choose.
            Both swift and go are popular and well paid but first of all I'm interested in learning blyadski facades and observers.
            Right now i'm shoes seller.
            Swift is a general-purpose programming language developed by Apple, for an operating system like iOS, macOS, watch OS, and Linux.
            Swift supports multiple programming paradigms and it is object-oriented, functional and imperative in nature.
            It is heavily designed to work with Appleâ€™s Cocoa and Cocoa Touch Frameworks.
            Go is a programming language also known by the name of Golang, created by Google in 2009.
            Go supports multi-paradigm like procedural, functional and concurrent.
            Its syntax is traditionally coming from C, but it has done a lot of modification to improve upon features like simplicity and safety.
            Tirka 23
          </div>
          {this.renderTreadMetaInfo()}
        </div>
      </Fragment>
    );
  };
  // action

  renderTreadAuthor = () => {
    const {
      author: {
        nickName = 'Ne Tot Pazan'
      }
    } = this.props;
    const [firstLetter, secondLetter] = nickName;

    return (
      <div className={styles.treadAuthor}>
        <div className={styles.treadAuthorBox}>
          {firstLetter + secondLetter}
        </div>
        <span>
         {nickName}
        </span>
      </div>
    );
  };

  renderTreadHeader = () => {
    const { treadView } = this.props;

    return (
      <div
        className={cx(styles.heading, {
          [styles.hovered]: treadView === PREVIEW
        })}
      >
        What to learn SWIFT or GO?
      </div>
    )
  };

  renderTreadMetaInfo = () => {
    const { subscribersCount, dislikesCount, likesCount } = this.props;

    return (
      <div className={styles.treadMetaInfo}>
        {subscribersCount} subscribers
        <span className={styles.divider}>-</span>
        {new Date().toISOString()}
        <span className={styles.divider}>-</span>
        <ThumbUpOutlinedIcon className={styles.icon} /> {likesCount}
        <span className={styles.divider}>-</span>
        <ThumbDownOutlinedIcon className={styles.icon} /> {dislikesCount}
      </div>
    );
  };

  renderPreviewTread = () => {
    const { answersCount } = this.props;

    return (
      <Fragment>
        <div>
          {this.renderTreadAuthor()}
          {this.renderTreadHeader()}
          {this.renderTreadMetaInfo()}
        </div>
        <div className={styles.answersCount}>
          <div className={styles.counterWrapper}>
            <QuestionAnswerOutlinedIcon />
            {answersCount}
          </div>
          <span>answer(s)</span>
        </div>
      </Fragment>
    );
  };

  render() {
    return (
      <div className={styles.wrapper}>
        {this.renderContent()}
      </div>
    );
  }
}

TreadComponent.propTypes = {
  heading: PropTypes.string,
  answersCount: PropTypes.number,
  subscribersCount: PropTypes.number,
  creationDate: PropTypes.object,
  content: PropTypes.string,
  likesCount: PropTypes.number,
  dislikesCount: PropTypes.number,
  treadView: PropTypes.oneOf(viewTypes),
  author: PropTypes.object,
  onLikeClick: PropTypes.func,
  onDislikeClick: PropTypes.func,
  onSubscribeClick: PropTypes.func,
  onEditClick: PropTypes.func
};

TreadComponent.defaultProps = {
  treadView: PREVIEW, // edit
  subscribersCount: randomize(),
  answersCount: randomize(),
  dislikesCount: randomize(),
  likesCount: randomize(),
  author: {}
};


