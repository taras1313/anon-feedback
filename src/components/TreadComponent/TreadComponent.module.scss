@import "../../styles/variables";

.wrapper {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  padding: 30px 0;
  border-bottom: 1px solid $dirty-white;
}

.treadAuthor {
  color: $black;
  font-weight: bold;
  display: flex;
  align-items: center;
  margin-bottom: 8px;
  font-size: 14px;
}

.treadAuthorBox {
  font-size: 12px;
  color: $white;
  display: flex;
  background: $dark-violet;
  box-shadow: -5px 5px $gray;
  justify-content: center;
  align-items: center;
  width: 26px;
  height: 26px;
  border-radius: 2px;
  text-transform: uppercase;
  margin-right: 8px;
}

.answersCount {
  display: flex;
  flex-direction: column;
  color: $light-green;
}

.heading {
  font-size: 21px;
  font-weight: bold;
}

.hovered {
  &:hover {
    color: $dark-violet;
    cursor: pointer;
  }
}

.description {
  margin: 16px 0;
  font-size: 15px;
  line-height: 24px;
}

.treadMetaInfo {
  color: $light-gray;
  display: flex;
  font-size: 13px;
  width: 100%;
  align-items: flex-end;
}

.counterWrapper {
  display: flex;
  justify-content: space-evenly;
  font-weight: bold;
  margin-bottom: 6px;
}

.divider {
  margin: 0 8px;
}

.icon {
  max-width: 20px;
  max-height: 20px;
  margin-right: 6px;
}