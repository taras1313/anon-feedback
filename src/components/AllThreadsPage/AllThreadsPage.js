import React from 'react';

import { ThreadComponent } from '../TreadComponent';
import Loader from '../Loader/Loader';

import Button from '@material-ui/core/Button';
import ButtonGroup from '@material-ui/core/ButtonGroup';
import WhatshotIcon from '@material-ui/icons/Whatshot';
import FavoriteIcon from '@material-ui/icons/Favorite';
import ScheduleIcon from '@material-ui/icons/Schedule'; 

import styles from './AllThreadsPage.module.scss';

const NEWEST = 'Newest';
const MOST_DISCUSSED = 'Most Discussed';
const MOST_LIKED = 'Most Liked';

const sortOptions = [
  { label: NEWEST, icon: <ScheduleIcon /> },
  { label: MOST_LIKED, icon: <FavoriteIcon /> },
  { label: MOST_DISCUSSED, icon: <WhatshotIcon /> }
];

export class AllThreadsPage extends React.Component {
  state = {
    sortBy: NEWEST,
    sortDirection: 'DESC'
  };

  sortByChangeHandler = () => ({ currentTarget: { name } }) => {
    const { sortBy, sortDirection } = this.state;

    if (name === sortBy) {
      const newDirection = sortDirection === 'ASC' ? 'DESC' : 'ASC';

      this.setState({ sortDirection: newDirection });
    } else {
      this.setState({ sortBy: name, sortDirection: 'DESC' });
    }
  };

  getSortedThreads = () => {
    const { sortBy } = this.state;
    const { threads } = this.props;

    switch (sortBy) {
      case NEWEST: {
        return [...threads].sort((a, b) => {
          return new Date(b.createdDate) - new Date(a.createdDate);
        });
      }

      case MOST_LIKED: {
        return [...threads].sort((a, b) => {
          return b.likesCount - a.likesCount;
        });
      }

      case MOST_DISCUSSED: {
        return [...threads].sort((a, b) => {
          return b.commentsCount - a.commentsCount;
        });
      }

      default: {
        return threads;
      }
    }
  };

  renderThreads() {
    const { sortDirection } = this.state;
    const { userId } = this.props;
    const sortedThreads = this.getSortedThreads().map(el => (
      <ThreadComponent userId={userId} key={el._id} threadView="preview" thread={el} />
    ));

    return sortedThreads;
    // return sortDirection === 'DESC' ? sortedThreads : sortedThreads.reverse();
  }

  renderButtons() {
    // const { sortDirection } = this.state;

    return sortOptions.map(({ label, icon }) => (
      <Button
        key={label}
        onClick={this.sortByChangeHandler()}
        name={label}
        className={this.isActive(label) ? styles.active : ''}
      >
        <div className={styles.iconWrapper}>{icon}</div>
        {label}
        {/*<div className={styles.iconWrapper}>*/}
        {/*{this.isActive(label) ? (*/}
        {/*sortDirection === 'DESC' ? (*/}
        {/*<ArrowDownwardIcon />*/}
        {/*) : (*/}
        {/*<ArrowUpwardIcon />*/}
        {/*)*/}
        {/*) : null}*/}
        {/*</div>*/}
      </Button>
    ));
  }

  componentDidMount() {
    const {
      actions: { getAllThreads }
    } = this.props;
    getAllThreads();
  }

  isActive = label => this.state.sortBy === label;

  render() {
    const { threads } = this.props;

    return (
      <>
        <div className={styles.buttonContainer}>
          <ButtonGroup size="small" aria-label="small outlined button group" color="primary">
            {this.renderButtons()}
          </ButtonGroup>
        </div>
        {threads.length ? this.renderThreads() : <Loader />}
      </>
    );
  }
}
