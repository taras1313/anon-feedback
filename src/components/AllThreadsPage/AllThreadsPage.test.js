import React from "react";
import { shallow } from "enzyme";
import { AllThreadsPage } from "./AllThreadsPage";
import { ThreadComponent } from "../TreadComponent";
import Button from "@material-ui/core/Button";

describe("AllThreadsPage", () => {
  const props = {
    threads: [
      {
        name: "dsfsf",
        _id: 1,
        likesCount: 1,
        commentsCount: 1,
        createdDate: "11/20/2015",
      },
      {
        name: "asdad",
        _id: 2,
        likesCount: 2,
        commentsCount: 0,
        createdDate: "13/20/2015",
      },
    ],
    actions: { getAllThreads: () => {} },
  };

  test("snapshot", () => {
    const wrapper = shallow(<AllThreadsPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("sort by newest", () => {
    const wrapper = shallow(<AllThreadsPage {...props} />);
    wrapper
      .find(Button)
      .at(1)
      .props()
      .onClick({ currentTarget: { name: 'Most Liked' } });
    console.log(wrapper.find(ThreadComponent).at(0).props());
    expect(wrapper.find(ThreadComponent).at(0).props().thread._id).toBe(2);
  });
});
