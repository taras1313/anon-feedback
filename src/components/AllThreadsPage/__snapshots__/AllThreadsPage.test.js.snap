// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllThreadsPage snapshot 1`] = `
<Fragment>
  <div
    className="buttonContainer"
  >
    <WithStyles(ForwardRef(ButtonGroup))
      aria-label="small outlined button group"
      color="primary"
      size="small"
    >
      <WithStyles(ForwardRef(Button))
        className="active"
        key="Newest"
        name="Newest"
        onClick={[Function]}
      >
        <div
          className="iconWrapper"
        >
          <ScheduleIcon />
        </div>
        Newest
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className=""
        key="Most Liked"
        name="Most Liked"
        onClick={[Function]}
      >
        <div
          className="iconWrapper"
        >
          <FavoriteIcon />
        </div>
        Most Liked
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className=""
        key="Most Discussed"
        name="Most Discussed"
        onClick={[Function]}
      >
        <div
          className="iconWrapper"
        >
          <WhatshotIcon />
        </div>
        Most Discussed
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ButtonGroup))>
  </div>
  <ThreadComponent
    actions={
      Object {
        "subscribeToThread": [Function],
        "unsubscribeFromThread": [Function],
      }
    }
    answersCount={4}
    author={Object {}}
    dislikeComment={[Function]}
    dislikeThread={[Function]}
    dislikesCount={0}
    key="1"
    likeComment={[Function]}
    likeThread={[Function]}
    likesCount={7}
    subscribersCount={9}
    thread={
      Object {
        "_id": 1,
        "commentsCount": 1,
        "createdDate": "11/20/2015",
        "likesCount": 1,
        "name": "dsfsf",
      }
    }
    threadView="preview"
  />
  <ThreadComponent
    actions={
      Object {
        "subscribeToThread": [Function],
        "unsubscribeFromThread": [Function],
      }
    }
    answersCount={4}
    author={Object {}}
    dislikeComment={[Function]}
    dislikeThread={[Function]}
    dislikesCount={0}
    key="2"
    likeComment={[Function]}
    likeThread={[Function]}
    likesCount={7}
    subscribersCount={9}
    thread={
      Object {
        "_id": 2,
        "commentsCount": 0,
        "createdDate": "13/20/2015",
        "likesCount": 2,
        "name": "asdad",
      }
    }
    threadView="preview"
  />
</Fragment>
`;
